src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java

```diff
@@ -981,249 +683,288 @@ public InputStream retrieveMetadata(String pid, String formatId)
+
+                    // Begin deletion process
+                    updateRefsFile(pid, absCidRefsPath, "remove");
+                    if (Files.size(absCidRefsPath) == 0) {
+                        Path objRealPath = getHashStoreDataObjectPath(pid);
+                        deleteList.add(FileHashStoreUtility.renamePathForDeletion(objRealPath));
+                        deleteList.add(FileHashStoreUtility.renamePathForDeletion(absCidRefsPath));
+                    } else {
+                        String warnMsg = "cid referenced by pid: " + pid
+                            + " is not empty (refs exist for cid). Skipping object deletion.";
+                        logFileHashStore.warn(warnMsg);
+                    }
+                    deleteList.add(FileHashStoreUtility.renamePathForDeletion(absPidRefsPath));
+                    // Delete all related/relevant items with the least amount of delay
+                    FileHashStoreUtility.deleteListItems(deleteList);
+                    deleteMetadata(pid);
+                    logFileHashStore.info("Data file and references deleted for: " + pid);
+
+                } finally {
+                    // Release lock
+                    releaseReferencedLockedCids(cid);
+                }
 
-            // Confirm that the object called to delete does not have a cid reference file
-            Path absCidRefsPath = getRealPath(cid, "refs", "cid");
-            if (Files.exists(absCidRefsPath)) {
-                // The cid is referenced by pids, do not delete.
-                return;
+            } catch (OrphanPidRefsFileException oprfe) {
+                // `findObject` throws this exception when the cid refs file doesn't exist,
+                // so we only need to delete the pid refs file and related metadata documents
+                Path absPidRefsPath = getHashStoreRefsPath(pid, HashStoreIdTypes.pid.getName());
+                deleteList.add(FileHashStoreUtility.renamePathForDeletion(absPidRefsPath));
+                // Delete items
+                FileHashStoreUtility.deleteListItems(deleteList);
+                deleteMetadata(pid);
+                String warnMsg = "Cid refs file does not exist for pid: " + pid
+                    + ". Deleted orphan pid refs file and metadata.";
+                logFileHashStore.warn(warnMsg);
+
+            } catch (OrphanRefsFilesException orfe) {
+                // `findObject` throws this exception when the pid and cid refs file exists,
+                // but the actual object being referenced by the pid does not exist
+                Path absPidRefsPath = getHashStoreRefsPath(pid, HashStoreIdTypes.pid.getName());
+                String cidRead = new String(Files.readAllBytes(absPidRefsPath));
+                // Since we must access the cid reference file, the `cid` must be synchronized
+                synchronizeReferencedLockedCids(cidRead);
```

Comment:
Is there any reason we need this extra lock? What is difference between `cid` and `cidRead`?

Reply:
In this scenario, the exception is thrown by `findObject`, so we never obtained a `cid` value. To proceed with the deletion routine without a `cid`, we must read the `cidRead` associated with the pid refs file (since only this exists), and finish the deletion routine with respect to the related `cidRead`.